File: src/sort/Haskell/Sort/BubbleSort.hs
module Sort.BubbleSort where

-- バブルソート
-- 一番大きい数値から順に、添え字の老い方へ移動する。
bubblesort :: Ord a => [a] -> [a]
bubblesort [] = []
bubblesort xs = let (xs',x) = splitAt (length xs - 1) $ swap xs
                in  bubblesort xs' ++ x
  where
    swap []         = []
    swap [x]        = [x]
    swap (x0:x1:xs) = if x0 < x1 then x0:swap (x1:xs) else x1:swap (x0:xs)

-- 一番小さい数値から順に、添え字の若い方へ移動する。
bubblesort2 :: Ord a => [a] -> [a]
bubblesort2 [] = []
bubblesort2 xs = let (x:xs') = foldr swap [] xs
                 in  x:bubblesort2 xs'
  where
    swap x []      = [x]
    swap x (y0:ys) = if y0 < x then y0:x:ys else x:y0:ys

File: src/Julia/bubblesort.jl
# バブルソート
function bubblesort!(xs::Array{T}) where T
    for i = (1:length(xs)-1)
        for j = (1:length(xs)-i)
            if xs[j] > xs[j+1]
                tmp = xs[j]
                xs[j] = xs[j+1]
                xs[j+1] = tmp
            end
        end
    end
end

function bubblesort(xs::Array{T}) where T
    xscopy = [x for x in xs]
    bubblesort!(xscopy)
    return xscopy
end

# 再帰版バブルソート
function bubblesortr!(xs::Array{T}) where T
    _bubblesortr!(xs, length(xs))
end

function bubblesortr(xs::Array{T}) where T
    xscopy = [x for x in xs]
    bubblesortr!(xscopy)
    return xscopy
end

function _bubblesortr!(xs::Array{T}, n) where T
    if n <= 1
        return
    end

    for i = (1:n-1)
        if xs[i] > xs[i+1]
            tmp = xs[i]
            xs[i] = xs[i+1]
            xs[i+1] = tmp
        end
    end

    _bubblesortr!(xs,n-1)
end
